### PR TITLE
#2413

### PR DESCRIPTION
**Pull request by BraveLittleTapok**
Created tests for android FileManager

### CODE REVIEW

File: jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/init/MobileAppFactory.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    public static MobileAppUIElement $(@MarkupLocator String locator, Object parent) {
+        return element(locator).setup(j->j.setParent(parent));
+    }
+    public static MobileAppUIElement $(@MarkupLocator By locator) {
+        return element(locator);
+    }
+    public static MobileAppUIElement $(WebElement webElement) {
+        return new MobileAppUIElement(webElement);
+    }
+    public static MobileAppUIElement $(@MarkupLocator By locator, Object parent) {
```
Comment (mf-code) [inline jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/init/MobileAppFactory.java:33]: Codacy found an issue: [The static method name '$' doesn't match '[a-z][a-zA-Z0-9]*'](https://app.codacy.com/gh/jdi-testing/jdi-light/pullRequest?prid=6361799)

**Comment by mf-code** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 27
           

Complexity increasing per file
==============================
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/common/app/Button.java  1
- jdi-light-mobile-tests/src/test/java/nativeapp_ios/SettingsAppTestInit.java  1
- jdi-light-mobile-tests/src/main/java/io/github/epam/enums/Navigation.java  1
- jdi-light-mobile-tests/src/test/java/nativeapp/ContactsTests.java  1
- jdi-light-mobile-tests/src/test/java/nativeapp_android/tests/CheckboxTests.java  1
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/base/MobileUIElement.java  4
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/common/MobileKeyboard.java  1
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/common/app/ios/InfoButton.java  1
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/common/MobileDevice.java  4
- jdi-light-mobile-tests/src/test/java/io/github/epam/mobile/tests/RootTests.java  1
- jdi-light-mobile-tests/src/main/java/io/github/epam/sections/Header.java  3
- jdi-light-mobile-tests/src/test/java/nativeapp_android/tests/AlarmAppTests.java  1
- jdi-light-mobile-tests/src/test/java/nativeapp_ios/tests/RemindersAppTests.java  1
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/asserts/ToggleButtonAssert.java  3
- jdi-light-mobile-tests/src/test/java/nativeapp_ios/MessagesAppTestsInit.java  2
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/common/app/ios/SystemButton.java  1
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/composite/ios/app/SegmentedControl.java  1
- jdi-light-mobile-tests/src/test/java/nativeapp_android/tests/SettingsAppTests.java  1
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/pageobjects/annotations/WebAnnotationsUtil.java  18
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/common/app/ios/AddContactButton.java  1
- jdi-light-mobile-tests/src/test/java/nativeapp_ios/tests/CalendarAppTests.java  1
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/common/app/ios/PickerWheel.java  1
- jdi-light-mobile-tests/src/test/java/io/github/epam/mobile/tests/StaticTestsInit.java  1
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/common/app/ios/Slider.java  1
- jdi-light-mobile-tests/src/test/java/nativeapp_android/MobileFileManagerInit.java  2
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/common/app/ios/Switch.java  2
- jdi-light-mobile-tests/src/test/java/nativeapp_ios/tests/SettingsAppTests.java  1
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/common/app/ios/DetailsDisclosureButton.java  1
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/MobileUtils.java  7
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/base/MobileBaseElement.java  2
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/composite/AndroidScreen.java  1
- jdi-light-mobile-tests/src/test/java/io/github/epam/mobile/tests/MobileRootTests.java  1
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/common/app/android/Checkbox.java  2
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/common/app/android/Switch.java  2
- jdi-light-mobile-tests/src/test/java/nativeapp_ios/CalendarAppTestsInit.java  2
- jdi-light-mobile-tests/src/test/java/nativeapp_android/ClockAppTestsInit.java  1
- jdi-light-mobile-tests/src/test/java/nativeapp_ios/tests/ContactsAppTests.java  1
- jdi-light-mobile-tests/src/test/java/nativeapp_ios/RemindersAppTestsInit.java  2
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/composite/ios/app/Menu.java  1
- jdi-light-mobile-tests/src/test/java/nativeapp_android/SettingsAppTestsInit.java  1
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/base/MobileAppBaseElement.java  2
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/composite/ios/app/SearchBar.java  1
- jdi-light-mobile-tests/src/test/java/nativeapp_android/tests/MobileFileManagerTests.java  3
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/composite/ios/app/NavigationBar.java  1
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/common/Icon.java  1
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/composite/MobileScreen.java  5
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/init/MobileAppFactory.java  3
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/asserts/SegmentedControlAssert.java  1
- jdi-light-mobile-tests/src/main/java/io/github/epam/custom/SearchResult.java  1
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/common/app/ios/TextField.java  2
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/common/MobileFileManager.java  4
- jdi-light-mobile-tests/src/test/java/nativeapp_android/tests/ToggleButtonTests.java  1
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/common/app/android/ToggleButton.java  2
- jdi-light-mobile-tests/src/main/java/nativeapp/contacts/CreateNewContact.java  1
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/asserts/CheckboxAssert.java  3
- jdi-light-mobile-tests/src/main/java/nativeapp/contacts/ContactsMainPage.java  1
- jdi-light-mobile-tests/src/test/java/nativeapp/elements/common/TextTests.java  1
- jdi-light-mobile-tests/src/test/java/nativeapp_ios/tests/MessagesAppTests.java  1
- jdi-light-mobile-tests/src/test/java/nativeapp_android/ApiDemosTestInit.java  1
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/asserts/SwitchAssert.java  3
- jdi-light-mobile-tests/src/main/java/io/github/epam/entities/Users.java  1
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/composite/ios/app/EditMenu.java  1
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/base/MobileAppUIElement.java  3
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/common/Text.java  1
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/CoordinateConversionHelper.java  1
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/MobileContextHolder.java  5
- jdi-light-mobile-tests/src/test/java/io/github/epam/steps/Preconditions.java  4
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/init/MobileUIFactory.java  3
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/common/AppManager.java  1
- jdi-light-mobile-tests/src/test/java/nativeapp_ios/ContactsAppTestsInit.java  1
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/common/app/android/Widget.java  1
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/settings/MobileSettings.java  2
         

Clones added
============
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/MobileUtils.java  2
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/common/app/android/Switch.java  1
- jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/common/app/android/ToggleButton.java  1
- jdi-light-mobile-tests/src/main/java/io/github/epam/StaticSite.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/jdi-testing/jdi-light/pullRequest?prid=6361799&bid=20648167)

File: jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/MobileUtils.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+                return (HasTouchActions) btnObj;
+        }
+        return GET_DEFAULT_BUTTON.execute(obj, buttonName);
+    };
+
+    private static String toButton(String buttonName) {
+        return buttonName.toLowerCase().contains("button") ? buttonName : buttonName + "button";
+    }
+
+    public static JFunc2<Object, String, HasTouchActions> GET_DEFAULT_APP_MENU_ITEM =
```
Comment (mf-code) [inline jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/MobileUtils.java:85]: Codacy found an issue: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/gh/jdi-testing/jdi-light/pullRequest?prid=6361799)

File: jdi-light-mobile-tests/src/main/java/io/github/epam/custom/Result.java [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package io.github.epam.custom;
+
+import com.epam.jdi.tools.DataClass;
+
+public class Result extends DataClass<Result> {
+    public String name, description, link;
```
Comment (mf-code) [inline jdi-light-mobile-tests/src/main/java/io/github/epam/custom/Result.java:6]: Codacy found an issue: [Use one line for each declaration, it enhances code readability.](https://app.codacy.com/gh/jdi-testing/jdi-light/pullRequest?prid=6361799)

File: jdi-light-mobile-tests/src/test/java/nativeapp_android/MobileFileManagerInit.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+import static nativeapp.android.AndroidFileManager.*;
+
+public class MobileFileManagerInit {
+
+    protected byte[] createdFileByte;
+    protected File createdFile;
+    protected File pulledFile;
+    public static final String FILE_NAME_TEST = "TEST.txt";
+    protected static final String PATH_TO_FILE_ANDROID = "/sdcard/download/" + FILE_NAME_TEST;
+    protected static final String PATH_TO_ANDROID_FOLDER = "/sdcard/download";
```
Comment (mf-code) [inline jdi-light-mobile-tests/src/test/java/nativeapp_android/MobileFileManagerInit.java:26]: Codacy found an issue: [Do not hardcode /sdcard.](https://app.codacy.com/gh/jdi-testing/jdi-light/pullRequest?prid=6361799)

File: jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/MobileUtils.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+            (obj, menuItemName) -> $(By.xpath(String.format("//XCUIElementTypeMenuItem[@name='%s']", menuItemName)),
+                    obj).setName(menuItemName);
+
+    public static JFunc2<Object, String, HasTouchActions> GET_APP_MENU_ITEM = (obj, buttonName) -> {
```
Comment (mf-code) [inline jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/MobileUtils.java:89]: Codacy found an issue: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/gh/jdi-testing/jdi-light/pullRequest?prid=6361799)

File: jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/MobileUtils.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+            throw exception("Cannot use this method. The driver needs to extend/implement the " + clazz.getName());
+        }
+    }
+
+
+    public static JFunc2<Object, String, HasTouchActions> GET_DEFAULT_BUTTON =
+            (obj, buttonName) -> $(By.xpath(String.format("//XCUIElementTypeButton[@name='%s']", buttonName)),
+                    obj).setName(buttonName);
+
+    public static JFunc2<Object, String, HasTouchActions> GET_BUTTON = (obj, buttonName) -> {
```
Comment (mf-code) [inline jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/MobileUtils.java:57]: Codacy found an issue: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/gh/jdi-testing/jdi-light/pullRequest?prid=6361799)

File: jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/init/MobileUIFactory.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    public static MobileUIElement $(@MarkupLocator String locator, Object parent) {
+        return element(locator).setup(j->j.setParent(parent));
+    }
+    public static MobileUIElement $(@MarkupLocator By locator) {
+        return element(locator);
+    }
+    public static MobileUIElement $(WebElement webElement) {
+        return new MobileUIElement(webElement);
+    }
+    public static MobileUIElement $(@MarkupLocator By locator, Object parent) {
```
Comment (mf-code) [inline jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/init/MobileUIFactory.java:33]: Codacy found an issue: [The static method name '$' doesn't match '[a-z][a-zA-Z0-9]*'](https://app.codacy.com/gh/jdi-testing/jdi-light/pullRequest?prid=6361799)

File: jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/CoordinateConversionHelper.java [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+        return getCoordinatesInViewport(x, y).moveBy(xOffset, yOffset);
+    }
+    public static Point getCoordinatesOnWebPage(Point point) {
+        return getCoordinatesOnWebPage(point.x, point.y);
+    }
+
+    /**
+     * The next two methods convert screen coordinates to viewport coordinates
+     */
+    public static Point getCoordinatesInViewport(int x, int y) {
```
Comment (mf-code) [inline jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/CoordinateConversionHelper.java:80]: Codacy found an issue: [Avoid reassigning parameters such as 'y'](https://app.codacy.com/gh/jdi-testing/jdi-light/pullRequest?prid=6361799)

File: jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/MobileUtils.java [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+        WebDriver driver = getDriver();
+        if (clazz.isInstance(driver)) {
+            return function.apply(clazz.cast(driver));
+        } else {
+            throw exception("Cannot use this method. The driver needs to extend/implement the " + clazz.getName());
+        }
+    }
+
+
+    public static JFunc2<Object, String, HasTouchActions> GET_DEFAULT_BUTTON =
```
Comment (mf-code) [inline jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/MobileUtils.java:53]: Codacy found an issue: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/gh/jdi-testing/jdi-light/pullRequest?prid=6361799)

File: jdi-light-mobile-tests/src/main/java/io/github/epam/entities/User.java [hunk 10]
```diff
@@ -0,0 +1,9 @@
+package io.github.epam.entities;
+
+import com.epam.jdi.tools.DataClass;
+
+public class User extends DataClass<User> {
+    public String name = "Roman", password = "Jdi1234", lastName,
```
Comment (mf-code) [inline jdi-light-mobile-tests/src/main/java/io/github/epam/entities/User.java:6]: Codacy found an issue: [Use one line for each declaration, it enhances code readability.](https://app.codacy.com/gh/jdi-testing/jdi-light/pullRequest?prid=6361799)

File: jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/init/MobileAppFactory.java [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+
+    public static MobileAppUIElement element(@MarkupLocator String locator) {
+        return locator.matches("[A-Z].*]")
+                ? new MobileAppUIElement().setName(locator)
+                : element(defineLocator(locator));
+    }
+    public static MobileAppUIElement element(@MarkupLocator By byLocator) {
+        return new MobileAppUIElement(byLocator);
+    }
+    public static MobileAppUIElement $(@MarkupLocator String locator) {
```
Comment (mf-code) [inline jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/init/MobileAppFactory.java:21]: Codacy found an issue: [The static method name '$' doesn't match '[a-z][a-zA-Z0-9]*'](https://app.codacy.com/gh/jdi-testing/jdi-light/pullRequest?prid=6361799)

File: jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/CoordinateConversionHelper.java [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+ * most of the methods return the browser coordinates. Thus this transformation
+ * is invoked when the touch actions are used
+ */
+public class CoordinateConversionHelper {
+
+    private static JavascriptExecutor js = (JavascriptExecutor) getDriver();
+    private static Function<WebDriver, Rectangle> TOOLBAR;
+    private static Function<WebDriver, Rectangle> WEBVIEW;
+    private static final String androidToolbarId = "com.android.chrome:id/toolbar";
+    private static final String iosToolbarXpath = "//XCUIElementTypeOther[@name='topBrowserBar']";
```
Comment (mf-code) [inline jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/CoordinateConversionHelper.java:26]: Codacy found an issue: [Avoid unused private fields such as 'iosToolbarXpath'.](https://app.codacy.com/gh/jdi-testing/jdi-light/pullRequest?prid=6361799)

File: jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/init/MobileUIFactory.java [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+    public static MobileUIElement $(@MarkupLocator String locator) {
+        return element(locator);
+    }
+    public static MobileUIElement $(@MarkupLocator String locator, Object parent) {
+        return element(locator).setup(j->j.setParent(parent));
+    }
+    public static MobileUIElement $(@MarkupLocator By locator) {
+        return element(locator);
+    }
+    public static MobileUIElement $(WebElement webElement) {
```
Comment (mf-code) [inline jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/init/MobileUIFactory.java:30]: Codacy found an issue: [The static method name '$' doesn't match '[a-z][a-zA-Z0-9]*'](https://app.codacy.com/gh/jdi-testing/jdi-light/pullRequest?prid=6361799)

File: jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/init/MobileUIFactory.java [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+                ? new MobileUIElement().setName(locator)
+                : element(defineLocator(locator));
+    }
+    public static MobileUIElement element(@MarkupLocator By byLocator) {
+        return new MobileUIElement(byLocator);
+    }
+    public static MobileUIElement $(@MarkupLocator String locator) {
+        return element(locator);
+    }
+    public static MobileUIElement $(@MarkupLocator String locator, Object parent) {
```
Comment (mf-code) [inline jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/init/MobileUIFactory.java:24]: Codacy found an issue: [The static method name '$' doesn't match '[a-z][a-zA-Z0-9]*'](https://app.codacy.com/gh/jdi-testing/jdi-light/pullRequest?prid=6361799)

File: jdi-light-mobile-tests/src/test/java/nativeapp_android/MobileFileManagerInit.java [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+import static com.epam.jdi.light.settings.WebSettings.logger;
+import static nativeapp.android.AndroidFileManager.*;
+
+public class MobileFileManagerInit {
+
+    protected byte[] createdFileByte;
+    protected File createdFile;
+    protected File pulledFile;
+    public static final String FILE_NAME_TEST = "TEST.txt";
+    protected static final String PATH_TO_FILE_ANDROID = "/sdcard/download/" + FILE_NAME_TEST;
```
Comment (mf-code) [inline jdi-light-mobile-tests/src/test/java/nativeapp_android/MobileFileManagerInit.java:25]: Codacy found an issue: [Do not hardcode /sdcard.](https://app.codacy.com/gh/jdi-testing/jdi-light/pullRequest?prid=6361799)

File: jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/init/MobileAppFactory.java [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+        return element(locator);
+    }
+    public static MobileAppUIElement $(@MarkupLocator String locator, Object parent) {
+        return element(locator).setup(j->j.setParent(parent));
+    }
+    public static MobileAppUIElement $(@MarkupLocator By locator) {
```
Comment (mf-code) [inline jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/init/MobileAppFactory.java:27]: Codacy found an issue: [The static method name '$' doesn't match '[a-z][a-zA-Z0-9]*'](https://app.codacy.com/gh/jdi-testing/jdi-light/pullRequest?prid=6361799)

File: jdi-light/src/main/java/com/epam/jdi/light/elements/init/PageFactory.java [hunk 17]
```diff
@@ -197,10 +198,10 @@ private static void initWebPage(WebPage webPage) {
     }
     //endregion
 
-    public static List<Class<?>> STOP_INIT_CLASSES = asList(
-        Object.class, WebPage.class, PageObject.class, UIElement.class,
+    public static ArrayList<Class<?>> STOP_INIT_CLASSES = new ArrayList<>(asList(
```
Comment (mf-code) [inline jdi-light/src/main/java/com/epam/jdi/light/elements/init/PageFactory.java:201]: Codacy found an issue: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/gh/jdi-testing/jdi-light/pullRequest?prid=6361799)

File: jdi-light-mobile-tests/src/main/java/io/github/epam/sections/LoginForm.java [hunk 18]
```diff
@@ -0,0 +1,10 @@
+package io.github.epam.sections;
+
+import com.epam.jdi.light.elements.composite.Form;
+import io.github.epam.entities.User;
+import org.openqa.selenium.WebElement;
+
+public class LoginForm extends Form<User> {
+	public WebElement name, password, loginButton;
```
Comment (mf-code) [inline jdi-light-mobile-tests/src/main/java/io/github/epam/sections/LoginForm.java:8]: Codacy found an issue: [Use one line for each declaration, it enhances code readability.](https://app.codacy.com/gh/jdi-testing/jdi-light/pullRequest?prid=6361799)

File: jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/CoordinateConversionHelper.java [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+            WEBVIEW = d -> {
+                Rectangle topToolbar = ((IOSDriver) d).findElementByAccessibilityId(iosToolbarAccessibilityId).getRect();
+                Rectangle bottomToolbar = ((IOSDriver) d).findElementByAccessibilityId(iosBottomToolbarAccessibilityId).getRect();
+                return new Rectangle(topToolbar.x, topToolbar.y + topToolbar.height, bottomToolbar.y - topToolbar.y - topToolbar.height, topToolbar.width);
+            };
+        }
+    }
+
+    @JDIAction(timeout = 5)
+    private static Rectangle toolbar() {
```
Comment (mf-code) [inline jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/CoordinateConversionHelper.java:49]: Codacy found an issue: [Avoid unused private methods such as 'toolbar()'.](https://app.codacy.com/gh/jdi-testing/jdi-light/pullRequest?prid=6361799)

File: jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/CoordinateConversionHelper.java [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+
+    private static JavascriptExecutor js = (JavascriptExecutor) getDriver();
+    private static Function<WebDriver, Rectangle> TOOLBAR;
+    private static Function<WebDriver, Rectangle> WEBVIEW;
+    private static final String androidToolbarId = "com.android.chrome:id/toolbar";
+    private static final String iosToolbarXpath = "//XCUIElementTypeOther[@name='topBrowserBar']";
+    private static final String iosToolbarAccessibilityId = "topBrowserBar";
+    private static final String androidWebviewClassName = "android.webkit.WebView";
+    private static final String iosBottomToolbarAccessibilityId = "BottomBrowserToolbar";
+    private static final String iosBottomToolbarXpath = "//XCUIElementTypeToolbar[@name='BottomBrowserToolbar']";
```
Comment (mf-code) [inline jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/CoordinateConversionHelper.java:30]: Codacy found an issue: [Avoid unused private fields such as 'iosBottomToolbarXpath'.](https://app.codacy.com/gh/jdi-testing/jdi-light/pullRequest?prid=6361799)

File: jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/init/MobileUIFactory.java [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+
+    public static MobileUIElement element(@MarkupLocator String locator) {
+        return locator.matches("[A-Z].*]")
+                ? new MobileUIElement().setName(locator)
+                : element(defineLocator(locator));
+    }
+    public static MobileUIElement element(@MarkupLocator By byLocator) {
+        return new MobileUIElement(byLocator);
+    }
+    public static MobileUIElement $(@MarkupLocator String locator) {
```
Comment (mf-code) [inline jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/init/MobileUIFactory.java:21]: Codacy found an issue: [The static method name '$' doesn't match '[a-z][a-zA-Z0-9]*'](https://app.codacy.com/gh/jdi-testing/jdi-light/pullRequest?prid=6361799)

File: jdi-light-mobile-tests/README.md [hunk 22]
```diff
@@ -0,0 +1,35 @@
+How to launch web tests on Android emulator  
+
+0) Download chrome driver 
+For more details visit: https://github.com/appium/appium/blob/master/docs/en/writing-running-appium/web/chromedriver.md
+It is recommended to use the version of chromedriver > 75. 
+The version of Chrome browser in the Android emulator must be the same.
+1) Set following settings in test.properties file: <br>
+<b>driver=android</b>; <br>
+<b>remote.type=appium</b>; <br>
+If your version of appium has its own URL (not http://0.0.0.0:4723), specify it in the <b>driver.remote.url</b> property
```
Comment (mf-code) [inline jdi-light-mobile-tests/README.md:10]: Codacy found an issue: [[no-literal-urls] Don’t use literal URLs without angle brackets](https://app.codacy.com/gh/jdi-testing/jdi-light/pullRequest?prid=6361799)

File: jdi-light-mobile-tests/README.md [hunk 23]
```diff
@@ -0,0 +1,35 @@
+How to launch web tests on Android emulator  
+
+0) Download chrome driver 
+For more details visit: https://github.com/appium/appium/blob/master/docs/en/writing-running-appium/web/chromedriver.md
```
Comment (mf-code) [inline jdi-light-mobile-tests/README.md:4]: Codacy found an issue: [[no-literal-urls] Don’t use literal URLs without angle brackets](https://app.codacy.com/gh/jdi-testing/jdi-light/pullRequest?prid=6361799)

File: jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/init/MobileAppFactory.java [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+                ? new MobileAppUIElement().setName(locator)
+                : element(defineLocator(locator));
+    }
+    public static MobileAppUIElement element(@MarkupLocator By byLocator) {
+        return new MobileAppUIElement(byLocator);
+    }
+    public static MobileAppUIElement $(@MarkupLocator String locator) {
+        return element(locator);
+    }
+    public static MobileAppUIElement $(@MarkupLocator String locator, Object parent) {
```
Comment (mf-code) [inline jdi-light-mobile/src/main/java/com/epam/jdi/light/mobile/elements/init/MobileAppFactory.java:24]: Codacy found an issue: [The static method name '$' doesn't match '[a-z][a-zA-Z0-9]*'](https://app.codacy.com/gh/jdi-testing/jdi-light/pullRequest?prid=6361799)

File: jdi-light-mobile-tests/src/test/java/io/github/epam/mobile/tests/MobileRootTests.java [hunk 25]
```diff
@@ -0,0 +1,33 @@
...
+import static io.github.epam.StaticSite.supportPage;
+import static io.github.epam.sections.Footer.aboutLink;
+import static io.github.epam.sections.Header.loginForm;
+import static io.github.epam.sections.Header.userIcon;
+import static io.github.epam.steps.Preconditions.shouldBeLoggedOut;
+import static io.github.epam.entities.Users.DEFAULT_USER;
+
+public class MobileRootTests extends StaticTestsInit {
+
+    private static final String nativeAppIos = "com.apple.Preferences";
```
Comment (mf-code) [inline jdi-light-mobile-tests/src/test/java/io/github/epam/mobile/tests/MobileRootTests.java:17]: Codacy found an issue: [Avoid unused private fields such as 'nativeAppIos'.](https://app.codacy.com/gh/jdi-testing/jdi-light/pullRequest?prid=6361799)

**Comment by mf-code** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- jdi-light-mobile-tests/src/test/java/nativeapp_android/tests/MobileFileManagerTests.java  3
- jdi-light-mobile-tests/src/test/java/nativeapp_android/MobileFileManagerInit.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/jdi-testing/jdi-light/pullRequest?prid=6361799&bid=20648167)

File: jdi-light-mobile-tests/src/test/java/nativeapp_android/MobileFileManagerInit.java (hunk 4 above)
Comment (mf-code) [inline jdi-light-mobile-tests/src/test/java/nativeapp_android/MobileFileManagerInit.java:26]: Codacy found an issue: [Do not hardcode /sdcard.](https://app.codacy.com/gh/jdi-testing/jdi-light/pullRequest?prid=6361799)

File: jdi-light-mobile-tests/src/test/java/nativeapp_android/MobileFileManagerInit.java (hunk 15 above)
Comment (mf-code) [inline jdi-light-mobile-tests/src/test/java/nativeapp_android/MobileFileManagerInit.java:25]: Codacy found an issue: [Do not hardcode /sdcard.](https://app.codacy.com/gh/jdi-testing/jdi-light/pullRequest?prid=6361799)

**Review by alexarm — APPROVED** [review]

**Pull request merged**